crypto/stark/src/autogen_sol.rs

```diff
@@ -0,0 +1,199 @@
+
+// The linter doesn't understand 'abstract' and thinks it's indendation
+
+// solhint-disable-next-line indent
+abstract contract RecurrenceTrace is DefaultConstraintSystem({}, {}, {}, {}) {{",
+        constraint_degree,
+        rows.len(),
+        n_cols,
+        blowup
+    );
+
+    // This specifies the lookup function
+    trace_layout_contract.push_str(
+        "\n    // This lets us map rows -> inverse index,
+    // In complex systems use a autogen binary search.
+    function row_to_offset(uint256 row) internal pure override returns(uint256) {",
+    );
+
+    // TODO  -this doesn't support negative rows
+    if rows.iter().enumerate().all(|(index, item)| index == *item) {
+        trace_layout_contract.push_str(
+            "     return row;
+         }",
+        );
+    } else {
+        trace_layout_contract.push_str(&binary_row_search_string(rows.as_slice()));
+        trace_layout_contract.push_str("}");
+    }
+
+    // This defines the trace layout function in solidity
+    trace_layout_contract.push_str("\n");
+    trace_layout_contract.push_str(&format!(
+        "
+    function layout_col_major() internal pure override returns(uint256[] memory) {{
+        uint256[] memory result = new uint256[]({});",
+        2 * trace_keys.len()
+    ));
+    for k in (0..2 * trace_keys.len()).step_by(2) {
+        let (i, j) = match trace_keys[k / 2] {
+            RationalExpression::Trace(i, j) => (i, j),
+            _ => panic!("Non trace rational expression in rows"),
+        };
+        trace_layout_contract.push_str(&format!(
+            "    (result[{}], result[{}]) = ({}, {});",
+            k,
+            k + 1,
+            i,
+            j
+        ));
+    }
+    trace_layout_contract.push_str(
+        "        return result;
+    }",
+    );
+
+    // This prints out the function called rows in solidity
+    trace_layout_contract.push_str(&format!(
+        "
+    function layout_rows() internal pure override returns(uint256[] memory) {{
+        uint256[] memory result = new uint256[]({});",
+        rows.len()
+    ));
+    for (index, row) in rows.iter().enumerate() {
+        trace_layout_contract.push_str(&format!("        result[{}] = {};", index, row));
+    }
+    trace_layout_contract.push_str(
+        "
+        return result;
+    }
+}
+    ",
+    );
+    trace_layout_contract
+}
+
+// TODO - This needs testing
+fn binary_row_search_string(rows: &[usize]) -> String {
+    if rows.len() == 1 {
+        return format!("return {}", rows[0]);
+    }
+    format!(
+        "
+    if (row > {}) {{
+        {}
+    }} else {{
+        {}
+    }}
+    return {};
+    ",
+        rows[rows.len() / 2],
+        binary_row_search_string(rows.get(0..(rows.len()) / 2).unwrap()),
+        binary_row_search_string(rows.get((rows.len() / 2)..rows.len()).unwrap()),
+        rows[rows.len() / 2]
+    )
+}
+
+fn lexicographic_compare(first: &RationalExpression, second: &RationalExpression) -> Ordering {
```

Comment:
Can we instead `derive(Ord)` on `RationalExpression`? Or `impl Ord on RationalExpression` ?

Reply:
It already exists, and that implementation won't work with this one.